### PR TITLE
feat: support cluster resource quota

### PR DESCRIPTION
**Pull request by zoumo**
#### What type of PR is this?

Feature: support tenant scope quota limit

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,51 @@
+
```
Comment (caohe) [inline .vscode/launch.json]: Do we need this vscode config file?

**Comment by lcw2** [comment]
can tanant access the API clusterresourcesquota？ 
I think tanant can access the API clusterresourcesquota to know  how much clusterresouces are used.

File: .vscode/launch.json (hunk 1 above)
Comment (zoumo) [reply .vscode/launch.json]: no, I will delete it

**Comment by zoumo** [comment]
> can tanant access the API clusterresourcesquota？ I think tanant can access the API clusterresourcesquota to know how much clusterresouces are used.

@lcw2 I think no, clusterresourcesquota is a system resource, should not be accessed by tenant by default, maybe we can expose it when system crd feature is ready @Silverglass

**Comment by lcw2** [comment]
ok,  when will this  system crd  feature  be release?

File: cmd/clusterresourcequota/install.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	quota "github.com/kubewharf/kubezoo/pkg/apis/quota"
```
Comment (Silverglass) [inline cmd/clusterresourcequota/install.go]: import style

File: docs/manually-setup-cn.md [hunk 3]
```diff
@@ -5,76 +5,21 @@
...
-NAME                            READY   AGE
-statefulset.apps/kubezoo        1/1     4h58m
-statefulset.apps/kubezoo-etcd   1/1     4h58m
-```
-
-### 暴露 KubeZoo service
-
-```console
-$ kubectl port-forward svc/kubezoo 6443:6443
+Export kubezoo server to 6443
```
Comment (Silverglass) [inline docs/manually-setup-cn.md:22]: 中文文档

File: hack/lib/gen_pki.sh [hunk 4]
```diff
@@ -19,15 +19,17 @@ set -xue
...
 CONTEXT_NAME=$(kubectl config current-context)
 if [ ${CONTEXT_NAME::5} != "kind-" ]; then
-     echo "Current kubectl context is not a kind cluster" >&2
-     exit 1
+    echo "Current kubectl context is not a kind cluster" >&2
+    exit 1
 fi
 KIND_SERVER="$(yq eval '.clusters.[]|select(.name=="'${CONTEXT_NAME}'")|.cluster.server' ~/.kube/config)"
-readonly TEMP_DIR=$REPO_ROOT/_output
+readonly TEMP_DIR=$(pwd -P)/_output/pki
```
Comment (Silverglass) [inline hack/lib/gen_pki.sh]: 用 $REPO_ROOT 比用 $(pwd -P) 更确定点？

File: hack/lib/gen_pki.sh [hunk 5]
```diff
@@ -51,8 +53,13 @@ gen_kubezoo_pki() {
     gen_kubernetes_cert
 }
 
-gen_ca(){
-    cat > $KUBEZOO_DIR/ca-config.json <<EOF
+gen_kubezoo_quota_pki() {
```
Comment (Silverglass) [inline hack/lib/gen_pki.sh]: 这个函数没用到？

File: pkg/clusterresourcequota/controllers/util.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
+// Copyright 2022 jim.zoumo@gmail.com
```
Comment (Silverglass) [inline pkg/clusterresourcequota/controllers/util.go]: Copyright 2022 The KubeZoo Authors.

**Review by Silverglass — CHANGES_REQUESTED** [review]

File: docs/manually-setup-cn.md (hunk 3 above)
Comment (zoumo) [reply docs/manually-setup-cn.md:22]: There is no problem here, this is the output of the script

File: docs/manually-setup.md [hunk 7]
```diff
@@ -6,79 +6,23 @@ a `kind` cluster as the upstream cluster.
...
-```
-
-### Export the KubeZoo service
-
-```console
-$ kubectl port-forward svc/kubezoo 6443:6443
+Export kubezoo server to 6443
 Forwarding from 127.0.0.1:6443 -> 6443
 Forwarding from [::1]:6443 -> 6443
 ```
```
Comment (Silverglass) [inline docs/manually-setup.md:28]: 下面创建 pod 时，需要指定 cpu 和 memory, 以便 resource quota  做认证；中文文档同理

File: cmd/kubezoo/app/server.go [hunk 8]
```diff
@@ -503,6 +507,30 @@ func buildProxyConfig(o *options.ProxyOptions) (*ProxyConfig, error) {
...
+					return nil, err
+				}
+				break
+			}
+		}
+		if clusterQuotaClient == nil {
+			klog.Warningf("upstream cluster does not have a resource 'clusterresourcequotas'")
+		}
+	} else {
+		klog.Warningf("failed to init cluster quota client with discovery error %v", err)
```
Comment (Silverglass) [inline cmd/kubezoo/app/server.go:533]: 也可能是 upstream 不存在这个 group, 所以这里应该判断下错误，而不是直接返回 error ?

**Review by Silverglass — APPROVED** [review]
Thank you very much for your contribution to tenant quota!

**Pull request merged**